clean_air/util/dataframe_converter.py

```diff
@@ -0,0 +1,148 @@
+import pandas as pd
+from geopandas import GeoDataFrame
+from shapely.geometry import Point
+import warnings
+
+
+def as_pandas_data(data, copy=True):
+    """
+    Sort iris cube style data into pandas style data.
+    :param data: Data array to be converted.
+    :param copy: Boolean value representing whether the data should be copied.
+    or not (this will return an error if the array is masked).
+    :return: Data array in pandas format.
+    """
+    if isinstance(data, np.ma.MaskedArray):
+        if not copy:
+            raise ValueError("Masked arrays must always be copied.")
+        data = data.astype('f').filled(np.nan)
+    elif copy:
+        data = data.copy()
+
+    return data
+
+
+def as_series(cube, index=None, copy=True):
+    """
+    Static function to convert an iris cube into a pandas series.
+
+    :param cube: iris cube (1-dimensional) for conversion.
+    :param index: 1-dimensional array representing the order of data for input
+    (if not supplied, the input order will be the same as the cube data).
+    :param copy: boolean value representing whether to keep a copy of the
+    data upon conversion.
+    :return: pandas Series containing information converted from iris cube.
+    """
+    # NOTE: This is not yet used in this converter, but it will be needed soon.
+    # It is as yet untested and unfinished; pulled from iris but in need of
+    # simplification and modification.
+    data = as_pandas_data(cube.data)
+    if not index:
+        if cube.dim_coords:
+            index = _as_pandas_coord(cube.dim_coords[0])
+    else:
+        index = _as_pandas_coord(index)
+
+    series = pd.Series(data, index)
+    if not copy:
+        _assert_shared(data, series)
+
+    return series
+
+
+def _make_geo(cube, x_coord, y_coord):
+    """
+    Helper function for conversion to geodataframe(s)
+    :param cube: cube for conversion
+    :param x_coord: iris dim coord representing x direction
+    :param y_coord: iris dim coord representing y direction
+    :return: list of geodataframes
+    """
+    dataframes = []
+    for sub_cube in cube.slices([x_coord, y_coord], ordered=True):
+        # Get the coordinates, data and geometry in the right
+        # order here:
+        x_coords_pd = []
+        y_coords_pd = []
+        data_pd = []
+        geom = []
+        i = 0
+        for x in x_coord.points:
+            # Reset the y-coord index to 0 each time we loop over the x-coord:
+            j = 0
+            for y in y_coord.points:
+                x_coords_pd.append(x)
+                y_coords_pd.append(y)
+                data_pd.append(sub_cube.data[i, j])
+                geom.append(Point(x, y))
+                # i and j represent the index values of the coords, so
+                # they need to stop counting at the end of the coord arrays.
+                if j < len(y_coord.points) - 1:
+                    j += 1
+            if i < len(x_coord.points) - 1:
+                i += 1
```

Comment:
Hmmm, I'm not sure about this.  I spent a long time trying to get this right and I'm still uneasy about it.  I will try your suggestions because I'm not happy about how messy this is turning out to be, but it might cause some problems.  I'll give it a go and let you know how I get on.

Reply:
But I believe that enumerate is what i tried to use first and just forgot the word, so thank you for the reminder.

Reply:
Well that works beautifully.  If only I'd remembered it in the first place.